src/scp/NominationProtocol.cpp

```diff
@@ -223,38 +223,61 @@ NominationProtocol::updateRoundLeaders()
     ZoneScoped;
     SCPQuorumSet myQSet = mSlot.getLocalNode()->getQuorumSet();
 
-    // initialize priority with value derived from self
-    std::set<NodeID> newRoundLeaders;
     auto localID = mSlot.getLocalNode()->getNodeID();
-    normalizeQSet(myQSet, &localID);
+    normalizeQSet(myQSet, &localID); // excludes self
 
-    newRoundLeaders.insert(localID);
-    uint64 topPriority = getNodePriority(localID, myQSet);
+    size_t maxLeaderCount = 1; // includes self
+    LocalNode::forAllNodes(myQSet,
+                           [&](NodeID const& cur) { ++maxLeaderCount; });
 
-    LocalNode::forAllNodes(myQSet, [&](NodeID const& cur) {
-        uint64 w = getNodePriority(cur, myQSet);
-        if (w > topPriority)
-        {
-            topPriority = w;
-            newRoundLeaders.clear();
-        }
-        if (w == topPriority && w > 0)
+    while (mRoundLeaders.size() < maxLeaderCount)
+    {
+        // initialize priority with value derived from self
+        std::set<NodeID> newRoundLeaders;
+
+        newRoundLeaders.insert(localID);
+        uint64 topPriority = getNodePriority(localID, myQSet);
+
+        LocalNode::forAllNodes(myQSet, [&](NodeID const& cur) {
+            uint64 w = getNodePriority(cur, myQSet);
```

Comment:
I know this was here before, but just to make sure: here we're using myQSet that intentionally does not include the local node? Wondering since the weight formula described in the whitepaper uses Q, and by definition Q includes self in all slices.

Reply:
Yes it's not exactly the same (and this whole thing needs to be reworked). We tweaked this because from what I remember, what is described in the paper sometimes yields no leader at all. This implementation always inserts `localID` in the `roundLeader` set, so we can safely reason about nodes outside of the local node.